src/include/meta.h

```diff
@@ -117,7 +117,10 @@ struct __wt_block_mods {
  * btree engine, and the block manager.
  */
 #define WT_CHECKPOINT "WiredTigerCheckpoint"
+#define WT_CHECKPOINT_MAX_LEN 40 /* WT_CHECKPOINT.INT64_MAX = 20+1+19 */
 #define WT_CKPT_FOREACH(ckptbase, ckpt) for ((ckpt) = (ckptbase); (ckpt)->name != NULL; ++(ckpt))
+#define WT_CKPT_FOREACH_NAME_OR_ORDER(ckptbase, ckpt) \
+    for ((ckpt) = (ckptbase); (ckpt)->name != NULL || (ckpt)->order != 0; ++(ckpt))
```

Comment:
Good question, at the moment we assign the order first and the name later, so we can potentially switch to using only the order. This seems to be true for the named checkpoints as well. I did not want to rely on that order assumption, which is easy to break in the future. If you don't have a strong reason to suggest otherwise, I will keep it as such for now.

Reply:
That's fine with me.  I just wanted to understand better.

Reply:
Would it make sense to rewrite `__wt_meta_checkpoint_free()` to not depend on anything having been set in the WT_CKPT structure, that is, it just clears out the structure? Then we don't need `WT_CKPT_FOREACH_NAME_OR_ORDER`, and we don't depend on any particular initialization order of the WT_CKPT structure. I think it's pretty easy, something like:
```
*** ../meta/meta_ckpt.c Sun May  9 18:00:50 2021
--- -   Sun May  9 18:07:58 2021 
***************  
*** 1188,1199 ****
      __wt_buf_free(session, &ckpt->addr);
      __wt_buf_free(session, &ckpt->raw); 
      __wt_free(session, ckpt->bpriv);
!     for (i = 0; i < WT_BLKINCR_MAX; ++i) {
!         blk_mod = &ckpt->backup_blocks[i];
!         __wt_buf_free(session, &blk_mod->bitstring);
!         __wt_free(session, blk_mod->id_str);
!         F_CLR(blk_mod, WT_BLOCK_MODS_VALID);
!     }   

      WT_CLEAR(*ckpt); /* Clear to prepare for re-use. */
  }   
--- 1188,1199 ----
      __wt_buf_free(session, &ckpt->addr);
      __wt_buf_free(session, &ckpt->raw); 
      __wt_free(session, ckpt->bpriv);
!     for (i = 0; i < WT_BLKINCR_MAX; ++i)
!         if ((blk_mod = &ckpt->backup_blocks[i]) != NULL) {
!           __wt_buf_free(session, &blk_mod->bitstring);
!           __wt_free(session, blk_mod->id_str);
!           F_CLR(blk_mod, WT_BLOCK_MODS_VALID);
!       }   

      WT_CLEAR(*ckpt); /* Clear to prepare for re-use. */
  }
```
might be enough. (And that resolves @keitharnoldsmith's comment as well.)

Reply:
I thought about it, `__wt_meta_checkpoint_free` wont be a problem, we can rewrite it to handle zeroed/freed checkpoints. `__wt_meta_ckptlist_free` is an issue here, it needs to walk over the whole list and call `__wt_meta_checkpoint_free` on individual checkpoints. Absent a check on `ckpt->name` or `ckpt->order`, we will not know when to terminate the list, which is an array of unknown size. Note: the size is not always `btree->ckpt_bytes_allocated`.

Reply:
Maybe it's time to have an "end-of-list" `WT_CKPT` field? Worrying about if/when fields are initialized is just asking for problems down the road, would we be having this discussion if it had been that way from the start?

Reply:
I totally agree, having a terminating field would make it simpler. I have created WT-7517 (and part of this project) to handle it. At this stage, I am trying to minimize the changes in this PR to get it committed.

Reply:
Marking resolved as will be handled with WT-7517